src/spatialdata/_core/query/relational_query.py

```diff
@@ -129,6 +147,255 @@ def _filter_table_by_elements(
+                    f"The element `{name}` is not annotated by the table. Skipping", UserWarning, stacklevel=2
+                )
+                continue
+
+    joined_table = table[joined_indices, :].copy() if joined_indices is not None else None
+
+    return element_dict, joined_table
+
+
+class JoinTypes(Enum):
+    """Available join types for matching elements to tables and vice versa."""
+
+    LEFT = left = partial(_left_join_spatialelement_table)
+    LEFT_EXCLUSIVE = left_exclusive = partial(_left_exclusive_join_spatialelement_table)
+    INNER = inner = partial(_inner_join_spatialelement_table)
+    RIGHT = right = partial(_right_join_spatialelement_table)
+    RIGHT_EXCLUSIVE = right_exclusive = partial(_right_exclusive_join_spatialelement_table)
+
+    def __call__(self, *args: Any) -> tuple[dict[str, Any], AnnData]:
+        return self.value(*args)
+
+    @classmethod
+    def get(cls, key: str) -> Callable[[dict[str, dict[str, Any]], AnnData], tuple[dict[str, Any], AnnData]] | None:
+        if key in cls.__members__:
+            return cls.__members__.get(key)
+        return None
+
+
+def join_sdata_spatialelement_table(
+    sdata: SpatialData, spatial_element_name: str | list[str], table_name: str, how: str = "LEFT"
+) -> tuple[dict[str, Any], AnnData]:
+    assert sdata.tables.get(table_name), f"No table with `{table_name}` exists in the SpatialData object."
+    table = sdata.tables[table_name]
+    if isinstance(spatial_element_name, str):
+        spatial_element_name = [spatial_element_name]
+
+    elements_dict: dict[str, dict[str, Any]] = {}
+    for name in spatial_element_name:
+        element_type, _, element = sdata._find_element(name)
+        elements_dict = _create_element_dict(element_type, name, element, elements_dict)
+    if "images" in elements_dict:
+        warnings.warn(
+            f"Images: `{', '.join(elements_dict['images'].keys())}` cannot be joined with a table",
+            UserWarning,
+            stacklevel=2,
+        )
+    if "tables" in elements_dict:
+        warnings.warn(
+            f"Tables: `{', '.join(elements_dict['tables'].keys())}` given in spatial_element_names cannot be "
+            f"joined with a table using this function.",
+            UserWarning,
+            stacklevel=2,
+        )
+
+    assert any(key in elements_dict for key in ["labels", "shapes", "points"]), (
+        "No valid element to join in spatial_element_name. Must provide at least one of either `labels`, `points` or "
+        "`shapes`."
+    )
+    if JoinTypes.get(how) is not None:
+        elements_dict, table = JoinTypes[how](elements_dict, table)
+    else:
+        raise ValueError(f"`{how}` is not a valid type of join.")
+
+    elements_dict = {
+        name: element for outer_key, dict_val in elements_dict.items() for name, element in dict_val.items()
+    }
+    return elements_dict, table
+
+
 def match_table_to_element(sdata: SpatialData, element_name: str) -> AnnData:
```

Comment:
This function and the tests associated to this function should be removed/adapted to use the join. In particular I think that the private function `_filter_table_by_elements()`, used internally by `match_table_to_element()` can be removed.

Reply:
Should we deprecate this function?

Reply:
Yes. I'd add a warning and call `match_element_to_table()` within it.

Reply:
the private function `_filter_table_by_elements` is still used in other parts of the code. Checking whether it can be removed would require more effort beyond the scope of this PR. `match_table_to_element` can use the join internally though.